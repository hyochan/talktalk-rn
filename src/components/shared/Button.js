// @flow
import React, { Component } from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
} from 'react-native';
import styled from 'styled-components/native';

import { ratio, colors } from '../../utils/Styles';

const StyledContainer = styled.View`
  flex: 1;
`;

const StyledButton = styled.View`
  background-color: ${colors.dodgerBlue};
  border-color: ${colors.dodgerBlue};
  border-radius: 4;
  border-width: 1;
  shadow-color: ${colors.dodgerBlue};
  shadow-radius: 4;
  shadow-opacity: 0.3;
  align-items: center;
  justify-content: center;
`;

const StyledButtonWhite = styled(StyledButton)`
  background-color: white;
  shadow-radius: 0;
  shadow-opacity: 0.0;
`;

const StyledButtonDisabled = styled.View`
  background-color: rgb(243,243,243);
  align-self: center;
  border-radius: 4;
  border-width: 2;
  border-color: #333;
  align-items: center;
  justify-content: center;
`;

const StyledText = styled.Text`
  font-size: 14;
  font-weight: bold;
  color: ${colors.dodgerBlue};
`;

const StyledTextWhite = styled(StyledText)`
  color: white;
`;

const StyledImageLeft = styled.Image`
  width: 24;
  height: 24;
  position: absolute;
  left: 16;
`;

type Props = {
  isWhite?: boolean;
  isLoading?: boolean;
  isDisabled?: boolean;
  onPress?: () => void;
  imgLeftSrc?: any;
  indicatorColor?: string;
  activeOpacity?: number;
  children?: any;
  width?: number;
  height?: number;
};

type State = {

};

class Button extends Component<Props, State> {
  static defaultProps: Props = {
    isWhite: false,
    isLoading: false,
    isDisabled: false,
    indicatorColor: 'white',
    activeOpacity: 0.5,
    width: 136,
    height: 60,
  };

  render() {
    if (this.props.isDisabled) {
      return (
        <StyledContainer>
          <StyledButtonDisabled>
            <StyledText>{this.props.children}</StyledText>
          </StyledButtonDisabled>
        </StyledContainer>
      );
    }
    if (this.props.isLoading) {
      return (
        <StyledContainer>
          <StyledButton>
            <ActivityIndicator size='small' color={this.props.indicatorColor} />
          </StyledButton>
        </StyledContainer>
      );
    }
    return (
      <StyledContainer>
        <TouchableOpacity
          activeOpacity={this.props.activeOpacity}
          onPress={this.props.onPress}
        >
          {
            !this.props.isWhite
              ? <StyledButton>
                {this.renderContent()}
              </StyledButton>
              : <StyledButtonWhite>
                {this.renderContent()}
              </StyledButtonWhite>
          }
        </TouchableOpacity>
      </StyledContainer>
    );
  }

  renderContent = () => {
    return <View style={{
      width: this.props.width,
      height: this.props.height,
      justifyContent: 'center',
      alignItems: 'center',
    }}>
      {
        this.props.imgLeftSrc
          ? <StyledImageLeft
            source={this.props.imgLeftSrc}
          />
          : null
      }
      { this.renderText() }
    </View>;
  }

  renderText = () => {
    return (
      this.props.isWhite
        ? <StyledText>{this.props.children}</StyledText>
        : <StyledTextWhite>{this.props.children}</StyledTextWhite>
    );
  }
}

export default Button;
