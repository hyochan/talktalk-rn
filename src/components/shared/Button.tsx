import React from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  ViewStyle,
  Image,
  Text,
  View,
} from 'react-native';
import styled from 'styled-components/native';

interface IStyledElement {
  white?: boolean;
};

interface IButtonContainer {
  width: number;
  height: number;
};

interface IProps {
  testID: string;
  containerStyle: ViewStyle;
  isWhite: boolean;
  isLoading: boolean;
  isDisabled: boolean;
  onPress: () => void;
  imgLeftSrc: any;
  indicatorColor: string;
  activeOpacity: number;
  children: any;
  width: number;
  height: number;
}

const StyledContainer = styled.View`
  flex: 1;
`;

const StyledButtonContainer = styled.View<IButtonContainer>`
  width: ${({ width }) => `${width}px`};
  height: ${({ height }) => `${height}px`};
  justify-content: center;
  align-items: center;
`;

const StyledButton = styled.View<IStyledElement>`
  background-color: ${({ white, theme: { colors: { dodgerBlue }}}) => white ? 'white' : dodgerBlue};
  border-color: ${({ theme: { colors: { dodgerBlue }}}) => dodgerBlue};
  border-radius: 4;
  border-width: 1;
  shadow-color: ${({ theme: { colors: { dodgerBlue }}}) => dodgerBlue};
  shadow-radius: ${({ white }) => white ? 0 : 4};
  shadow-opacity: ${({ white }) => white ? 0.0 : 0.3};
  align-items: center;
  justify-content: center;
`;

const StyledButtonDisabled = styled.View`
  background-color: rgb(243,243,243);
  align-self: center;
  border-radius: 4;
  border-width: 2;
  border-color: #333;
  align-items: center;
  justify-content: center;
`;

const StyledText = styled.Text<IStyledElement>`
  font-size: 14;
  font-weight: bold;
  color: ${({ white, theme: { colors: { dodgerBlue }}}) => white ? 'white' : dodgerBlue};
`;

const StyledImageLeft = styled.Image`
  width: 24;
  height: 24;
  position: absolute;
  left: 16;
`;

function Button({
  isDisabled,
  isWhite,
  activeOpacity,
  onPress,
  isLoading,
  width,
  height,
  indicatorColor,
  imgLeftSrc,
  children
}: Partial<IProps>) { 
  return (
    <StyledContainer>
      {isDisabled ? (
        <StyledButtonDisabled>
          <StyledText>{children}</StyledText>
        </StyledButtonDisabled>
      ) : (
        <TouchableOpacity
          testID="press_id"
          activeOpacity={activeOpacity}
          onPress={onPress}
        >
          <StyledButton white={isWhite}>
            <StyledButtonContainer
              width={width}
              height={height}
            >
              {isLoading
                ? <ActivityIndicator size='small' color={indicatorColor} />
                : (imgLeftSrc ? <StyledImageLeft source={imgLeftSrc} /> : undefined)
              }
              {!isLoading
                ? <StyledText white={!isWhite}>{children}</StyledText>
                : undefined
              }
            </StyledButtonContainer>
          </StyledButton>            
        </TouchableOpacity>
      )}      
    </StyledContainer>
  );
}

Button.defaultProps = {
  isWhite: false,
  isLoading: false,
  isDisabled: false,
  indicatorColor: 'white',
  activeOpacity: 0.5,
  width: 136,
  height: 60,
};

export default Button;
