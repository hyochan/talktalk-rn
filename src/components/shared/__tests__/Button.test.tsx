import 'react-native';
import * as React from 'react';
import Button from '../Button';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

let cnt = 0;
const onPress = () => {
  cnt++;
};
const component: React.ReactElement = <Button onPress={onPress}/>;

describe('[Button] rendering test', () => {
  it('renders [enabled] as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
  it('renders [disabled] as expected', () => {
    const disabledComponent: React.ReactElement = <Button isDisabled={true}/>;
    const json = renderer.create(disabledComponent).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[Button] interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let testingLib: any;

  beforeAll(() => {
    rendered = renderer.create(component);
    testingLib = render(component);
  });

  it('should fireEvent when peer image is pressed', () => {
    const touchBtn = testingLib.getByTestId('press_id');
    fireEvent(touchBtn, 'press');
    expect(cnt).toEqual(1);
  });
});
