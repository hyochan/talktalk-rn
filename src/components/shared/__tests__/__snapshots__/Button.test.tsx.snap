// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] rendering test renders [disabled] as expected 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "rgb(243,243,243)",
          "borderColor": "#333",
          "borderRadius": 4,
          "borderWidth": 2,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 14,
            "fontWeight": "bold",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`[Button] rendering test renders [enabled] as expected 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="press_id"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(58,139,255)",
            "borderColor": "rgb(58,139,255)",
            "borderRadius": 4,
            "borderWidth": 1,
            "justifyContent": "center",
            "shadowColor": "rgb(58,139,255)",
            "shadowOpacity": 0.3,
            "shadowRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 60,
            "justifyContent": "center",
            "width": 136,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 14,
                "fontWeight": "bold",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
